_rules/text-contrast-afw4f7.md

```diff
@@ -0,0 +1,366 @@
+---
+id: afw4f7
+name: Text has minimum contrast
+rule_type: atomic
+
+description: |
+  This rule checks that the highest possible contrast of every text character with its background meets the minimal contrast requirement
+
+accessibility_requirements:
+  wcag20:1.4.3: # Contrast (Minimum)
+    forConformance: true
+    failed: not satisfied
+    passed: further testing needed
+    inapplicable: further testing needed
+
+input_aspects:
+  - DOM Tree
+  - CSS Styling
+  - Language
+
+acknowledgements:
+  authors:
+    - Brian Bors
+    - Kasper Isager
+    - Wilco Fiers
+---
+
+## Applicability
+
+Any [visible](#visible) character in a [text node](https://dom.spec.whatwg.org/#text) that is a [child](https://dom.spec.whatwg.org/#concept-tree-child) (in the [flat tree](https://drafts.csswg.org/css-scoping/#flat-tree)) of an HTML element, except if the [text node](https://dom.spec.whatwg.org/#text) is a [descendant](https://dom.spec.whatwg.org/#concept-shadow-including-descendant) of an element that:
+
+- Has a [semantic role](#semantic-role) that inherits from [widget](https://www.w3.org/TR/wai-aria-1.1/#widget)
+- Is used in the [accessible name](#accessible-name) of a [widget](https://www.w3.org/TR/wai-aria-1.1/#widget) that is [disabled]()
+- Has a [semantic role](#semantic-role) of [group](https://www.w3.org/TR/wai-aria-1.1/#group) that is [disabled]()
+
+**Note**: When the text color is the same as the background color, the [text node](https://dom.spec.whatwg.org/#text) is not [visible](#visible), and so it does not need to be tested for contrast.
+
+## Expectation
+
+For each test target, the [highest possible contrast](#highest-possible-contrast) between the [foreground colors](#foreground-colors-of-text) and [background colors](#background-colors-of-text) is at least 4.5:1 or 3.0:1 for [larger scale text](#large-scale-text), except if the [text node](https://dom.spec.whatwg.org/#text) is [decorative](#decorative), or doesn't express anything in [human language](https://www.w3.org/TR/WCAG21/#dfn-human-language-s).
```

Comment:
This would still express something in human language. I don't think this is a concern.

Reply:
The text nodes would be "hel" and "lo". None of them express something in human language.

I see the second exception in the expectation as "except if the text node (in which the test target, i.e. the character is) doesn't express anything". 

Given that "hel" doesn't express anything in human language, each of the three characters would fall into that exception and all would pass.
That is,
```html
<span style="color: #FFE; background: #FFF;">hel</span>
```
is passing the rule (for each of the 3 test targets) because the text node does not express anything in human language. And similarly,
```html
<span style="color: #FFE; background: #FFF;">hel</span><span style="color: #FFD; background: #FFF;">lo</span>
```
passes the rule (for each of the 5 test targets) because none of the text nodes express anything in human language.

I do agree that it is not a big concern given that it is a edge case and will only result in passing a few extra things.